pkg/dns/ibm/private/dnssvcs_provider.go

```diff
@@ -0,0 +1,273 @@
+		record.Spec.RecordTTL = defaultDNSSVCSRecordTTL
+	}
+
+	for _, target := range record.Spec.Targets {
+		updated := false
+
+		listResult, response, err := p.dnsService.ListResourceRecords(listOpt)
+		if err != nil {
+			if response != nil && response.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("createOrUpdateDNSRecord: failed to list the dns record: %w", err)
+			}
+		}
+		if listResult == nil {
+			return fmt.Errorf("createOrUpdateDNSRecord: invalid result")
+		}
+
+		for _, resourceRecord := range listResult.ResourceRecords {
+
+			if *resourceRecord.Name == DNSName {
+				updated = true
+
+				updateOpt := p.dnsService.NewUpdateResourceRecordOptions(p.config.InstanceID, zone.ID, *resourceRecord.ID)
+				updateOpt.SetName(DNSName)
+				if record.Spec.RecordType == iov1.CNAMERecordType {
+					inputRData, error := p.dnsService.NewResourceRecordUpdateInputRdataRdataCnameRecord(target)
+					if error != nil {
+						return fmt.Errorf("createOrUpdateDNSRecord: failed to create CNAME inputRData for the dns record: %w", err)
+					}
+					updateOpt.SetRdata(inputRData)
+				} else {
+					inputRData, error := p.dnsService.NewResourceRecordUpdateInputRdataRdataARecord(target)
+					if error != nil {
+						return fmt.Errorf("createOrUpdateDNSRecord: failed to create A inputRData for the dns record: %w", err)
+					}
+					updateOpt.SetRdata(inputRData)
+				}
+				updateOpt.SetTTL(record.Spec.RecordTTL)
+				_, _, err := p.dnsService.UpdateResourceRecord(updateOpt)
+				if err != nil {
+					return fmt.Errorf("createOrUpdateDNSRecord: failed to update the dns record: %w", err)
+				}
+				log.Info("updated DNS record", "record", record.Spec, "zone", zone, "target", target)
+			}
+
+		}
+		if !updated {
+			createOpt := p.dnsService.NewCreateResourceRecordOptions(p.config.InstanceID, zone.ID)
+			createOpt.SetName(DNSName)
+			createOpt.SetType(string(record.Spec.RecordType))
+
+			if record.Spec.RecordType == iov1.CNAMERecordType {
+				inputRData, error := p.dnsService.NewResourceRecordInputRdataRdataCnameRecord(target)
+				if error != nil {
+					return fmt.Errorf("createOrUpdateDNSRecord: failed to create CNAME inputRData for the dns record: %w", err)
+				}
+				createOpt.SetRdata(inputRData)
+			} else {
+				inputRData, error := p.dnsService.NewResourceRecordInputRdataRdataARecord(target)
+				if error != nil {
+					return fmt.Errorf("createOrUpdateDNSRecord: failed to create A inputRData for the dns record: %w", err)
+				}
+				createOpt.SetRdata(inputRData)
+			}
+			createOpt.SetTTL(record.Spec.RecordTTL)
+			_, _, err := p.dnsService.CreateResourceRecord(createOpt)
+			if err != nil {
+				return fmt.Errorf("createOrUpdateDNSRecord: failed to create the dns record: %w", err)
+			}
+			log.Info("created DNS record", "record", record.Spec, "zone", zone, "target", target)
+		}
+	}
+
+	return nil
+}
+
+func validateInputDNSData(record *iov1.DNSRecord, zone configv1.DNSZone) error {
```

Comment:
*nit* does zone need to be pass-by-value here? Doesn't that do an unnecessary memory copy?
Same comment goes for `cis_provider.go` to be consistent.

Reply:
Done.